### ISSUE TITLE
more hash collisions after hash `Type` PR

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#49975]
After https://github.com/JuliaLang/julia/pull/49636 the following hash collision happens:

```julia
julia> using QuantumCumulants

julia> hf = FockSpace(:c)

julia> a = Destroy(hf,:a)

julia> typeof(a)
Destroy{FockSpace{Symbol}, Symbol, Int64, Nothing}

julia> typeof(a')
Create{FockSpace{Symbol}, Symbol, Int64, Nothing}

julia> hash(a)
0xc35039ec05f5ae3b

julia> hash(a')
0xc35039ec05f5ae3b

julia> hash(a) == hash(a')
true
```

It causes https://github.com/qojulia/QuantumCumulants.jl/blob/7a00cbbdbc3c702668689900f96303607ed65c6a/test/test_fock.jl#LL8C1-L13C34 to fail.

Now, this is not really a bug but it seems that the hashing should do better here to not cause collisions in this case where the types are obviously different.

**Comment by inkydragon** [comment]
https://github.com/qojulia/QuantumCumulants.jl/blob/7a00cbbdbc3c702668689900f96303607ed65c6a/src/fock.jl#L71-L73
```jl
for f in [:Destroy,:Create]
    @eval function Base.hash(op::T, h::UInt) where T<:($(f))
        hash($(f), hash(op.hilbert, hash(op.name, hash(op.aon, h))))
    end
end
```

It looks like they have a custom `Base.hash` function.

**Comment by KristofferC** [comment]
Thanks, that makes the repro smaller:

```
julia> hash(QuantumCumulants.Create)
0x77cfa1eef01bca90

julia> hash(QuantumCumulants.Destroy)
0x77cfa1eef01bca90
```

**Comment by gbaraldi** [comment]
Isn't that type piracy now?

**Comment by KristofferC** [comment]
Defining hash for an object? I don't think so... (Note they are not defining it for `Type`).

**Comment by gbaraldi** [comment]
Oddly I can't reproduce this locally on the latest master with asserts.

**Comment by KristofferC** [comment]
> Oddly I can't reproduce this locally on the latest master with asserts.

Yes, I noticed that too. Seems to only be when backported to 1.9. I guess something with the Type based hashing that ends up getting called has changed.

**Comment by KristofferC** [comment]
Since this seems fixed on 1.10 I guess we can close this (since the PR is no longer backported).
